### ISSUE TITLE
HODLR doesn't work with latest versions of Eigen

### ISSUE THREAD

**Issue by michael-hartmann** [sivaramambikasaran/HODLR#13]
I want to thank you for your work and for making the code publicly available. I am using your library heavily to evaluate determinants that are numerically out of reach for standard algorithms like Cholesky decomposition. Without your work my current studies would not be possible.

When working on my code, I recently updated my Eigen library. After the update, the HODLR library gave wrong results. I found that the problem is caused by Eigen versions starting from 3.2.8. Up to 3.2.7, however, everything works fine.

The reason is that in version 3.2.8 Eigen made a change in `FullPivLU.h`. Here is the relevant quote from the [changelog](http://eigen.tuxfamily.org/index.php?title=ChangeLog#Eigen_3.2.8):
> Make FullPivLU::solve use rank() instead of nonzeroPivots(). 

Here is a diff of the change that causes the HODLR library to break:
```
*** FullPivLU.h_327     2018-03-23 16:02:57.268313250 +0100
--- FullPivLU.h_328     2018-03-23 16:03:22.409744343 +0100
*************** struct solve_retval<FullPivLU<_MatrixTyp
*** 688,694 ****
       */  
  
      const Index rows = dec().rows(), cols = dec().cols(),
!               nonzero_pivots = dec().nonzeroPivots();
      eigen_assert(rhs().rows() == rows);
      const Index smalldim = (std::min)(rows, cols);
  
--- 688,694 ----
       */  
  
      const Index rows = dec().rows(), cols = dec().cols(),
!               nonzero_pivots = dec().rank();
      eigen_assert(rhs().rows() == rows);
      const Index smalldim = (std::min)(rows, cols);
```

Note that a single line causes the problem: The change from `nonzero_pivots = dec().nonzeroPivots();` to `nonzero_pivots = dec().rank();`.

The problem may be reproduced using the HODLR_Test program. Here is the output using Eigen 3.2.7
```
Number of particles is: 50000

Setting things up...
Time taken is: 0.001317

Assembling the matrix in HODLR form...
Time taken is: 0.140323

Exact matrix vector product...
Time taken is: 46.2392

Fast matrix matrix product...
Time taken is: 0.013503

Factoring the matrix...
Time taken is: 0.445196

Solving the system...
Time taken is: 0.016266

Error in computed solution: 2.79761e-09

Error in matrix matrix product: 9.35927e-11

Computing the log determinant...
Time taken is: 0.008733

Log determinant is: -15227.07956830352
```
and here is the output of Eigen 3.3 (note that the result is the same also with other versions greater than 3.2.7)
```
Number of particles is: 50000

Setting things up...
Time taken is: 0.001308

Assembling the matrix in HODLR form...
Time taken is: 0.139998

Exact matrix vector product...
Time taken is: 45.5071

Fast matrix matrix product...
Time taken is: 0.012069

Factoring the matrix...
Time taken is: 0.453063

Solving the system...
Time taken is: 0.015234

Error in computed solution: -nan

Error in matrix matrix product: 1.45292e-10

Computing the log determinant...
Time taken is: 0.007363

Log determinant is: nan
```

Unfortunately, I don't have enough oversight over the HODLR library or knowledge of C++ to further trace down the bug. If you decide only to support Eigen up to version 3.2.7, a comment in the readme would be helpful.

**Comment by michael-hartmann** [comment]
It seems that the latest version fixes this issue. :+1:

**Comment by sivaramambikasaran** [comment]
Fixed in #17
